File: src/main/scala/BettingOddsCalculator.scala
import scala.io.StdIn._
import scala.util.{Failure, Success, Try}

/**
 * BettingOddsCalculator is a simple utility for calculating implied probabilities
 * from various types of odds: decimal, moneyline, and fractional.
 */
object BettingOddsCalculator {

  /**
   * Checks if a given string represents a valid fraction (e.g., "1/2", "3/4").
   *
   * @param input The input string to check.
   * @return True if the input is a valid fraction, false otherwise.
   */
  private def isFraction(input: String): Boolean = {
    // Regular expression to match fractions (e.g., "1/2", "3/4")
    val fractionPattern = """^\d+/\d+$""".r
    fractionPattern.matches(input)
  }

  /**
   * Converts a fractional odds string into a decimal odds value.
   *
   * @param fraction The fractional odds as a string (e.g., "1/2").
   * @return The equivalent decimal odds as a Float.
   * @throws IllegalArgumentException if division by zero is detected.
   */
  private def fractionToDecimal(fraction: String): Float = {
    val parts = fraction.split("/")
    val numerator = parts(0).toFloat
    val denominator = parts(1).toFloat

    if (denominator != 0) {
      numerator / denominator
    } else {
      throw new IllegalArgumentException("Division by zero")
    }
  }

  /**
   * Calculates the implied probability from decimal odds entered by the user.
   */
  def decimalProbCalc(): Unit = {
    println(Console.BLUE + "Enter the Decimal Odds: ")

    val userEnteredDecimalValue: Try[Float] = Try(readFloat())

    userEnteredDecimalValue match {
      case Failure(_) =>
        println(Console.RED + "Please Enter a Numeric Value for Decimal Odds")

      case Success(value) =>
        println(Console.GREEN + "Calculating the Odds")
        val ans = ProbFunctions.decimalProb(value)
        println(Console.GREEN + s"The Implied Probability is $ans")
    }
  }

  /**
   * Calculates the implied probability from moneyline odds entered by the user.
   */
  def moneyProbCalc(): Unit = {
    println(Console.BLUE + "Enter the Moneyline: ")

    val userEnterMoneylineValue: Try[Float] = Try(readFloat())

    userEnterMoneylineValue match {
      case Failure(_) =>
        println(Console.RED + "Please Enter an Integer Value for Moneyline Odds")

      case Success(value) =>
        println(Console.GREEN + "Calculating the Odds")
        val ans = ProbFunctions.moneylineProb(value)
        println(Console.GREEN + s"The Implied Probability is $ans")
    }
  }

  /**
   * Calculates the implied probability from fractional odds entered by the user.
   */
  def fractionProbCalc(): Unit = {
    println(Console.BLUE + "Enter the Fractional Odds: ")

    val userEnterFractionValue: String = readLine()

    if (isFraction(userEnterFractionValue)) {
      println(Console.GREEN + "Calculating the Odds")

      val decimalConvertedValue: Try[Float] = Try(fractionToDecimal(userEnterFractionValue))

      decimalConvertedValue match {
        case Failure(_) =>
          println(Console.RED + "Error: Invalid fraction or division by zero.")

        case Success(value) =>
          val ans = ProbFunctions.fractionProb(value)
          println(Console.GREEN + s"The Implied Probability is $ans")
      }
    } else {
      println(Console.RED + "Enter Fractional Odds in a Proper Format")
    }
  }
}

File: src/bettingOddsCalculator.rs
use std::io;

mod betting_odds_calculator {

    fn fraction_to_decimal(fraction: &str) -> Result<f32, &'static str> {
        let parts: Vec<&str> = fraction.split('/').collect();
        let numerator: f32 = parts[0].parse().unwrap_or_default();
        let denominator: f32 = parts[1].parse().unwrap_or_default();

        if denominator != 0.0 {
            Ok(numerator / denominator)
        } else {
            Err("Division by zero")
        }
    }

    pub fn decimal_prob_calc() {
        println!("\x1b[34mEnter the Decimal Odds:");
        
        let mut user_entered_decimal_value = String::new();
        io::stdin().read_line(&mut user_entered_decimal_value)
            .expect("Failed to read input");
        
        let user_entered_decimal_value: f32 = match user_entered_decimal_value.trim().parse() {
            Ok(num) => num,
            Err(_) => {
                println!("\x1b[31mPlease Enter a Numeric Value for Decimal Odds");
                return;
            }
        };

        println!("\x1b[32mCalculating the Odds");
        let ans = prob_functions::decimal_prob(user_entered_decimal_value);
        println!("\x1b[32mThe Implied Probability is {}", ans);
    }

    pub fn money_prob_calc() {
        println!("\x1b[34mEnter the Moneyline:");

        let mut user_entered_moneyline_value = String::new();
        io::stdin().read_line(&mut user_entered_moneyline_value)
            .expect("Failed to read input");

        let user_entered_moneyline_value: f32 = match user_entered_moneyline_value.trim().parse() {
            Ok(num) => num,
            Err(_) => {
                println!("\x1b[31mPlease Enter an Integer Value for Moneyline Odds");
                return;
            }
        };

        println!("\x1b[32mCalculating the Odds");
        let ans = prob_functions::moneyline_prob(user_entered_moneyline_value);
        println!("\x1b[32mThe Implied Probability is {}", ans);
    }

    pub fn fraction_prob_calc() {
        println!("\x1b[34mEnter the Fractional Odds:");

        let mut user_entered_fraction_value = String::new();
        io::stdin().read_line(&mut user_entered_fraction_value)
            .expect("Failed to read input");

        let user_entered_fraction_value = user_entered_fraction_value.trim();

        if is_fraction(user_entered_fraction_value) {
            println!("\x1b[32mCalculating the Odds");

            match fraction_to_decimal(user_entered_fraction_value) {
                Ok(value) => {
                    let ans = prob_functions::fraction_prob(value);
                    println!("\x1b[32mThe Implied Probability is {}", ans);
                }
                Err(_) => println!("\x1b[31mError: Invalid fraction or division by zero."),
            }
        } else {
            println!("\x1b[31mEnter Fractional Odds in a Proper Format");
        }
    }

    fn is_fraction(input: &str) -> bool {
        let fraction_pattern = regex::Regex::new(r"^\d+/\d+$").unwrap();
        fraction_pattern.is_match(input)
    }
}

